File: 3464-maximize-the-distance-between-points-on-a-square/3464-maximize-the-distance-between-points-on-a-square.py
from typing import List

class Solution:
    def maxDistance(self, side: int, points: List[List[int]], k: int) -> int:
        # Step 1: Map 2D points to 1D positions along the perimeter
        # We also need to keep the original coordinates for Manhattan calculation
        perimeter_points = []
        for x, y in points:
            if y == 0:          # Bottom
                d = x
            elif x == side:     # Right
                d = side + y
            elif y == side:     # Top
                d = 2 * side + (side - x)
            else:               # Left
                d = 3 * side + (side - y)
            perimeter_points.append((d, x, y))
        
        # Sort by 1D position
        perimeter_points.sort()
        n = len(perimeter_points)

        def get_manhattan(p1, p2):
            return abs(p1[1] - p2[1]) + abs(p1[2] - p2[2])

        def check(mid_dist):
            # Try different starting points due to circular nature
            # Since k is small, we only need to try a few starting indices
            for start_idx in range(min(n, n // k + 1)):
                count = 1
                last_pt = perimeter_points[start_idx]
                
                for j in range(start_idx + 1, n):
                    # Validate against the current 'mid_dist'
                    if get_manhattan(last_pt, perimeter_points[j]) >= mid_dist:
                        # Additionally, check distance to the VERY FIRST point 
                        # to maintain the minimum distance across the wrap-around
                        if count == k - 1:
                            if get_manhattan(perimeter_points[j], perimeter_points[start_idx]) >= mid_dist:
                                return True
                        else:
                            count += 1
                            last_pt = perimeter_points[j]
            return False

        # Step 2: Binary Search on the possible answer
        low = 1
        high = 2 * side # Max Manhattan distance possible in a square
        ans = 1
        
        while low <= high:
            mid = (low + high) // 2
            if check(mid):
                ans = mid
                low = mid + 1
            else:
                high = mid - 1
                
        return ans

File: 3464-maximize-the-distance-between-points-on-a-square/3464-maximize-the-distance-between-points-on-a-square.java
class Solution {
    record Point(int x, int y, long unRolledDist) {
        public static int dist(Point a, Point b) {
            return Math.abs(a.x - b.x) + Math.abs(a.y - b.y);
        }
    }

    /*
     * Unfold the points into a 1D array
     * Sort the array
     * Binary search on res with range [1, side], as there are at least 4 points, the best score can be size.
     *
     * Now for some dist,
     * we need to select k points from the array, where each point has at least dist distance to the previous point.
     * */
    public int maxDistance(int side, int[][] points, int k) {
        Point[] unfoldedArr = new Point[points.length];
        int arrLen = 0;

        for (int[] point : points) {
            int x = point[0];
            int y = point[1];
            long unRolledDist = 0;
            if (x == 0) unRolledDist = y;
            else if (y == side) unRolledDist = (long) side + x;
            else if (x == side) unRolledDist = 3L * side - y;
            else if (y == 0) unRolledDist = 4L * side - x;
            unfoldedArr[arrLen++] = new Point(x, y, unRolledDist);
        }
        Arrays.sort(unfoldedArr, Comparator.comparingLong(Point::unRolledDist));

        int left = 1;
        int right = side;
        while (left < right) {
            int mid = (left + right + 1) >> 1;
            if (isPossible(unfoldedArr, mid, k)) left = mid;
            else right = mid - 1;
        }
        return left;
    }

    private boolean isPossible(Point[] unfoldedArr, int limit, int k) {
        int n = unfoldedArr.length;
        for (int i = 0; i < n; i++)
            if (isPossible(unfoldedArr, i, limit, k)) return true;

        return false;
    }

    private boolean isPossible(Point[] unfoldedArr, int firstIdx, int limit, int k) {
        int n = unfoldedArr.length;
        int prevIdx = firstIdx;
        Point prev = unfoldedArr[prevIdx];
        k--;

        while (k > 0 && prevIdx < n - 1) {
            int nextIdx = lowerBound(unfoldedArr, prev, prevIdx + 1, n - 1, limit);
            if (nextIdx == -1 || (k == 1 && Point.dist(unfoldedArr[firstIdx], unfoldedArr[nextIdx]) < limit)) return false;
            prevIdx = nextIdx;
            prev = unfoldedArr[prevIdx];
            k--;
        }
        return k == 0;
    }

    // finds first point whose dist from prev is as least limit
    private int lowerBound(Point[] unfoldedArr, Point prev, int left, int right, int limit) {
        while (left < right) {
            int mid = (left + right) >> 1;
            if (Point.dist(prev, unfoldedArr[mid]) >= limit) right = mid;
            else left = mid + 1;
        }
        return Point.dist(prev, unfoldedArr[left]) < limit ? -1 : left;
    }
}
